### ISSUE TITLE
how to deploy rootfs on flash

### ISSUE THREAD

**Issue by fanguannan0706** [OP-TEE/optee_os#680]
Hi,
I used repo kit to flash the optee on Hikey board, and I found the rootfs is not really deployed on the flash(emmc), So I can't save what I write. I'd like to know what can I do ?

**Comment by jbech-linaro** [comment]
Hi @fanguannan0706 ,

You must update the rootfs and re-trigger the updates to the boot image.
https://github.com/OP-TEE/build/blob/master/hikey.mk#L277
https://github.com/OP-TEE/build/blob/master/hikey.mk#L330

`make boot-img` should take care of that and all the dependencies. But that you for some reason cannot get that to work, then please do a `make all` instead.

If you issue is that you own built files etc cannot be found in the device, then it's probably because you haven't point them out to be included in the root fs. That is typically what is done here:
https://github.com/OP-TEE/build/blob/master/hikey.mk#L273-L275

Those lines are doing some find, sed magic, so it's probably easier to see how it should be done by looking at the qemu.mk instead:
https://github.com/OP-TEE/build/blob/master/qemu.mk#L129-L148

// Joakim

**Comment by fanguannan0706** [comment]
Actually what I mean is that I find there is only a tmpfs on Hikey, is there any way to flash a filesystem on emmc?

**Comment by vchong** [comment]
You can flash a rootfs to the system partition: `fastboot flash system <rootfs img>` and modify the initramfs here to pivot to that rootfs. I'm not sure how to do that though.

Might be easier to follow this guide - https://github.com/96boards/documentation/wiki/HiKeyoptee - if you want, but it's a bit outdated. If you're using Release 2.0.0 and newer (as of Apr 2, 2016), skip the linux, optee_linuxdriver, and modprobe related parts.

**Comment by vchong** [comment]
Sorry, the instructions on https://github.com/96boards/documentation/wiki/HiKeyoptee will not work for now on Release 2.0.0 and later due to kernel incompatibility. If you want to use it, you'll have to wait till it gets updated. No ETA for now.

**Comment by fanguannan0706** [comment]
Ok, thank you very much. I'll wait for the update, at the same time, I will try to flash a rootfs.

**Comment by fanguannan0706** [comment]
I copy optee_armtz.ko and optee.ko to the file system, but the modules' version magic are incompatible. So I try to change the host kernel version to "3.18.0-linaro-hikey" and recompile it. Then when I try to insmod the modules there is some new error and dmesg is "Unknown symbol _mcount (err 0)".

**Comment by vchong** [comment]
It's mentioned above that the instructions currently will NOT work for Release 2.0.0. Please wait for it to be updated.

**Comment by jbech-linaro** [comment]
Not that the "fastboot flash system" way, is a bit different, in our simple setup that isn't used. But having that said you could use it if you're building for Android for for Debian. Both has been verified OK by different people.
Android: https://github.com/d3zd3z/optee_android_manifest (to be merged to offical Linaro gits any day)
Debian: https://github.com/OP-TEE/optee_os#552-debian-based--96boards-rpb-experimental

**Comment by jbech-optee-adm** [comment]
Question has been answered, but If the issue persists or you have more questions, please re-open the issue.

**Comment by fanguannan0706** [comment]
I succeed to flash Debian on hikey and used uart3 as console before, but now, I can't get any output from uart3. I tried to change it to uart0, it can't work either. And I tried to re-flash the system as guide some times, but both(uart0 and uart3) can't work. I'm not sure what I do cause this mistake, it confused me so much. I have checked usb-to-ttl port and Dupont line, they are no problem. What should I check to find the problem? Is there some way to display via HDMI?
